### ISSUE TITLE
False `InvalidTemplateParam`: Eloquent `Collection::map()`/`mapWithKeys()` with a non-Model callback

### ISSUE THREAD

**Issue by alies-dev** [psalm/laravel-psalm-plugin#1029]
## Summary

The conditional return type on `Eloquent\Collection::map()` (and `mapWithKeys()`) added in #548 reintroduces an `InvalidTemplateParam` error whenever the callback returns a non-`Model` value. Root cause is an upstream Psalm limitation: a conditional return `(T is SomeClass ? A : B)` is **not narrowed** when `T` is a scalar that provably cannot be `SomeClass` — Psalm keeps the union `A|B` and bound-checks both branches.

## Current stub

`stubs/common/Database/Eloquent/Collection.phpstub`:

```php
/**
 * @template TMapValue
 * @param  callable(TModel, TKey): TMapValue  $callback
 * @psalm-return (TMapValue is \Illuminate\Database\Eloquent\Model ? static<TKey, TMapValue> : \Illuminate\Support\Collection<TKey, TMapValue>)
 */
public function map(callable $callback) {}
```

`mapWithKeys()` below has the identical shape and the same flaw.

## Reproduction (non-Model map → false positive)

```php
<?php declare(strict_types=1);

use Illuminate\Database\Eloquent\Collection;
use Illuminate\Database\Eloquent\Model;

final class Plan extends Model {}

/**
 * @param Collection<int, Plan> $plans
 * @return list<non-falsy-string>
 */
function nicknames(Collection $plans): array
{
    return $plans
        ->map(static fn(Plan $plan): string => 'plan-' . $plan->getKey())
        ->filter()
        ->unique()
        ->values()
        ->all();
}
```

```
InvalidTemplateParam: Extended template param TModel of Illuminate\Database\Eloquent\Collection<int, non-falsy-string>&static expects type Illuminate\Database\Eloquent\Model, type non-falsy-string given (see https://psalm.dev/183)
```

Reproduction conditions: the collection must be parameterised by a **concrete** model; the callback must produce a type narrower than `string` (e.g. `non-falsy-string`); and a downstream call on the result (e.g. `->filter()`) is needed to force materialisation of the union member's template params.

## Underlying Psalm defect

A collection-independent probe shows the conditional is not narrowed for scalars:

```php
/**
 * @template T
 * @param T $value
 * @psalm-return (T is \Illuminate\Database\Eloquent\Model ? 'is-model' : 'not-model')
 */
function classify(mixed $value): string { return 'not-model'; }

function probe(string $s): void {
    $r = classify($s);
    /** @psalm-trace $r */ // Trace: $r: 'is-model'|'not-model'
}
```

`string` can never be a `Model`, yet Psalm yields the union `'is-model'|'not-model'` instead of `'not-model'`. In `map()` the `'is-model'` analogue is `static<TKey, TMapValue>` = `Collection<…, TModel of Model>` parameterised with a non-`Model` `TMapValue`, which violates the bound. The `is Model` guard does not exclude the branch.

## Recommended plugin fix (workaround)

Until Psalm narrows the dead branch, drop the conditional and return the base collection for both methods:

```php
/** @psalm-return \Illuminate\Support\Collection<TKey, TMapValue> */
public function map(callable $callback) {}

/** @psalm-return \Illuminate\Support\Collection<TMapWithKeysKey, TMapWithKeysValue> */
public function mapWithKeys(callable $callback) {}
```

This is sound (`Eloquent\Collection` is a `Support\Collection`) and produces zero false positives. Verified against several real call sites: callbacks returning `Model`, concrete model (`Tag`/`Masterclass`), and non-`Model` (`non-falsy-string`) values all type-check, and the declared `Support\Collection<int, ConcreteModel>` returns match exactly.

### Rejected alternatives

- `static<TKey, TMapValue>` (current) — false positive for non-`Model` callbacks (this issue).
- `static<TKey, \Illuminate\Database\Eloquent\Model>` — passes the non-`Model` case but **drops `TMapValue`**: `map(fn(Tag): Tag)` then infers `Collection<int, Model>`, more general than a declared `Support\Collection<int, Tag>` → `MoreSpecificReturnType` / `LessSpecificReturnStatement`.
- `static<TKey, TMapValue&Model>` — Psalm bound-checks the raw substituted `TMapValue` and ignores the intersection narrowing; error persists.

Tradeoff: the workaround loses the `Eloquent\Collection` subtype for the all-`Model` case. That matches base `Collection::map` and the runtime `toBase()` fallback, and is the price of the Psalm limitation. The precise conditional can be restored once Psalm narrows `scalar is ClassName` to `false` (worth a separate vimeo/psalm issue).

## Environment

- psalm/plugin-laravel: `master` (`5.72.0-2-g3203d06fc68`)
- vimeo/psalm: `7.0.0-beta19`
- PHP: `8.4.21`
- Introduced by: #548 (fix for #500)

**Comment by alies-dev** [comment]
Upstream Psalm root cause filed: vimeo/psalm#11867 (conditional return true branch `self<T>`/`static<T>` bound-checked without applying the `T is X` guard). Once that lands, the precise `static<TKey, TMapValue>` conditional can be restored; until then the plain `Support\Collection<TKey, TMapValue>` workaround above avoids the false positives.

**Comment by alies-dev** [comment]
Converted to the repository internal ideas list (`.alias/issues.md`) so GitHub issues stay focused on active work. Reopen or recreate this when it is prioritized.
